grails-app/controllers/org/pih/warehouse/inventory/InventoryController.groovy

```diff
@@ -575,49 +575,70 @@ class InventoryController {
+
+        List<InventoryItem> inventoryItems = dashboardService.getExpiredStock(categorySelected, location)
+        List<Category> categories = inventoryItems?.collect { it.product.category }?.unique()
+
+        List<Map> data
+        if (withBinLocation) {
+            data = productAvailabilityService.getQuantityOnHandByBinLocation(location)
+        } else {
+            data = productAvailabilityService.getQuantityOnHandByInventoryItem(location)
+                    .collect{ key, val -> [ inventoryItem: key, quantity: val ] }
         }
+        data = data.findAll { inventoryItems.contains(it.inventoryItem) }
+
         if (params.format == "csv") {
             def filename = "Expired stock | " + location?.name + ".csv"
             response.setHeader("Content-disposition", "attachment; filename=\"${filename}\"")
-            render(contentType: "text/csv", text: getCsvForInventoryMap(expiredStockMap, [:]))
+            render(contentType: "text/csv", text: getCsvForInventoryMap(data, withBinLocation))
             return
         }
 
-        [inventoryItems: inventoryItems, quantityMap: quantityMap, categories: categories, categorySelected: categorySelected]
+        [
+                data: data,
+                categories: categories,
+                categorySelected: categorySelected,
+        ]
     }
 
 
     def listExpiringStock() {
-        def expirationStatus = params.status
-        def location = Location.get(session.warehouse.id)
-        def category = (params.category) ? Category.get(params.category) : null
-        def inventoryItems = dashboardService.getExpiringStock(category, location, expirationStatus)
-        def categories = inventoryItems?.collect { it?.product?.category }?.unique().sort {
+        String expirationStatus = params.status
+        Location location = Location.get(session.warehouse.id)
+        Category category = Category.get(params.category)
+        Boolean withBinLocation = params.boolean("withBinLocation")
+
+        List<InventoryItem> inventoryItems = dashboardService.getExpiringStock(category, location, expirationStatus)
+        List<Category> categories = inventoryItems?.collect { it?.product?.category }?.unique().sort {
             it.name
         }
-        def quantityMap = productAvailabilityService.getQuantityOnHandByInventoryItem(location)
-        def expiringStockMap = [:]
-        inventoryItems.each { inventoryItem ->
-            expiringStockMap[inventoryItem] = quantityMap[inventoryItem]
+
+        List<Map> data
+        if (withBinLocation) {
+            data = productAvailabilityService.getQuantityOnHandByBinLocation(location)
+        } else {
+            data = productAvailabilityService.getQuantityOnHandByInventoryItem(location)
+                    .collect{ key, val -> [ inventoryItem: key, quantity: val ] }
         }
+        data = data.findAll { inventoryItems.contains(it.inventoryItem) }
```

Comment:
this seems to be a repetition of what is done in the `listExpiredStock`. Maybe we could create a private method in this controller (as you know me - I'd prefer to have that type of logic in the service, but no must, since this probably needs to be refactored anyway, anytime soon).